src/Core/Resolvers/SqlQueryEngine.cs

```diff
@@ -233,8 +248,87 @@ private OkObjectResult FormatFindResult(JsonDocument jsonDoc, FindRequestContext
                                   path,
                                   nvc: context!.ParsedQueryString,
                                   after);
-            rootEnumerated = rootEnumerated.Append(nextLink);
-            return OkResponse(JsonSerializer.SerializeToElement(rootEnumerated));
+
+            rootEnumerated = RemoveExtraFieldsInResponseWithMultipleItems(rootEnumerated.ToList(), extraFieldsInResponse);
+            return OkResponse(JsonSerializer.SerializeToElement(rootEnumerated.Append(nextLink)));
+        }
+
+        /// <summary>
+        /// To support pagination and $first clause with Find requests, it is necessary to provide the nextLink
+        /// field in the response. For the calculation of nextLink, the fields such as primary keys, fields in $orderby clause
+        /// are retrieved from the database in addition to the fields requested in the $select clause.
+        /// However, these fields are not required in the response.
+        /// This function helps to determine those additional fields that are present in the response.
+        /// </summary>
+        /// <param name="response">Response json retrieved from the database</param>
+        /// <param name="context">FindRequestContext for the GET request.</param>
+        /// <returns>Additional fields that are present in the response</returns>
+        private static IEnumerable<string> DetermineExtraFieldsInResponse(JsonElement response, FindRequestContext context)
+        {
+            // context.FieldsToBeReturned will contain the fields requested in the $select clause.
+            // If $select clause is absent, it will contain the list of columns that can be returned in the
+            // response taking into account the include and exclude fields configured for the entity.
+            HashSet<string> fieldsToBeReturned = new(context.FieldsToBeReturned);
+
+            HashSet<string> fieldsPresentInResponse = new();
+
+            if (response.ValueKind is not JsonValueKind.Array)
+            {
+                foreach (JsonProperty property in response.EnumerateObject())
+                {
+                    fieldsPresentInResponse.Add(property.Name);
+                }
+            }
+            else
+            {
+                foreach (JsonProperty property in response.EnumerateArray().Last().EnumerateObject())
+                {
+                    fieldsPresentInResponse.Add(property.Name);
+                }
+            }
+
+            return fieldsPresentInResponse.Except(fieldsToBeReturned);
```

Comment:
```suggestion
            return fieldsPresentInResponse.Except(context.FieldsToBeReturned);
``` 

would directly using `context.FieldsToBeReturned` allow you to avoid the extra hashset allocation?